File: backend/src/main/scala/domains/message/objects/request/MarkConversationReadRequest.scala
package domains.message.objects.request

import domains.message.objects.*
import io.circe.{Decoder, DecodingFailure, Encoder}
import io.circe.generic.semiauto.deriveEncoder

/** 标记会话已读请求体，message 模式必须携带 messageId。 */
final case class MarkConversationReadRequest(
  mode: MarkConversationReadMode,
  messageId: Option[MessageId]
)

/** 提供已读请求体 codec，并在解码阶段校验 mode/messageId 的组合。 */
object MarkConversationReadRequest:
  given Encoder[MarkConversationReadRequest] = deriveEncoder[MarkConversationReadRequest]

  given Decoder[MarkConversationReadRequest] = Decoder.instance { cursor =>
    for
      mode <- cursor.downField("mode").as[MarkConversationReadMode]
      messageId <- cursor.downField("messageId").as[Option[MessageId]]
      _ <- mode match
        case MarkConversationReadMode.Message if messageId.isEmpty =>
          Left(DecodingFailure("messageId is required when mode is message", cursor.history))
        case _ =>
          Right(())
    yield MarkConversationReadRequest(
      mode = mode,
      messageId = mode match
        case MarkConversationReadMode.Conversation => None
        case MarkConversationReadMode.Message => messageId
    )
  }

File: frontend/src/objects/message/request/MarkConversationReadRequest.ts
import type { MessageId } from '@/objects/message/MessageId'
import type { MarkConversationReadMode } from '@/objects/message/request/MarkConversationReadMode'

/** 标记会话已读请求；按模式决定是否需要提供消息游标。 */
export type MarkConversationReadRequest =
  | {
      mode: Extract<MarkConversationReadMode, 'conversation'>
    }
  | {
      mode: Extract<MarkConversationReadMode, 'message'>
      messageId: MessageId
    }
